source/socket.c

```diff
@@ -0,0 +1,263 @@
+    return socket->vtable->socket_write_fn(socket, cursor, written_fn, user_data);
+}
+
+int aws_socket_get_error(struct aws_socket *socket) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_get_error_fn);
+    return socket->vtable->socket_get_error_fn(socket);
+}
+
+bool aws_socket_is_open(struct aws_socket *socket) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_is_open_fn);
+    return socket->vtable->socket_is_open_fn(socket);
+}
+
+static enum aws_socket_impl_type aws_socket_get_default_impl_type(void);
+static int aws_socket_impl_type_validate_platform(enum aws_socket_impl_type type);
+int aws_socket_init(struct aws_socket *socket, struct aws_allocator *alloc, const struct aws_socket_options *options) {
+
+    // 1. get socket type & validate type is avliable the platform
+    enum aws_socket_impl_type type = options->impl_type;
+    if (type == AWS_SOCKET_IMPL_PLATFORM_DEFAULT) {
+        type = aws_socket_get_default_impl_type();
+    }
+
+    if (aws_socket_impl_type_validate_platform(type)) {
+        AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "Invalid event loop type on the platform.");
+        return aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
+    }
+
+    // 2. setup vtable based on socket type
+    switch (type) {
+        case AWS_SOCKET_IMPL_POSIX:
+            return aws_socket_init_posix(socket, alloc, options);
+        case AWS_SOCKET_IMPL_WINSOCK:
+            return aws_socket_init_winsock(socket, alloc, options);
+        case AWS_SOCKET_IMPL_APPLE_NETWORK_FRAMEWORK:
+            // Apple Network Framework is not implemented yet. We should not use it yet.
+            AWS_ASSERT(false && "Invalid socket implementation on platform.");
+            return aws_socket_init_apple_nw_socket(socket, alloc, options);
+        default:
+            AWS_ASSERT(false && "Invalid socket implementation on platform.");
+            return aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
+    }
+}
+
+int aws_socket_get_bound_address(const struct aws_socket *socket, struct aws_socket_endpoint *out_address) {
+    if (socket->local_endpoint.address[0] == 0) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p fd=%d: Socket has no local address. Socket must be bound first.",
+            (void *)socket,
+            socket->io_handle.data.fd);
+        return aws_raise_error(AWS_IO_SOCKET_ILLEGAL_OPERATION_FOR_STATE);
+    }
+    *out_address = socket->local_endpoint;
+    return AWS_OP_SUCCESS;
+}
+
+void aws_socket_endpoint_init_local_address_for_test(struct aws_socket_endpoint *endpoint) {
+    (void)endpoint;
+    struct aws_uuid uuid;
+    AWS_FATAL_ASSERT(aws_uuid_init(&uuid) == AWS_OP_SUCCESS);
+    char uuid_str[AWS_UUID_STR_LEN] = {0};
+    struct aws_byte_buf uuid_buf = aws_byte_buf_from_empty_array(uuid_str, sizeof(uuid_str));
+    AWS_FATAL_ASSERT(aws_uuid_to_str(&uuid, &uuid_buf) == AWS_OP_SUCCESS);
+
+#if defined(AWS_ENABLE_KQUEUE) || defined(AWS_ENABLE_EPOLL)
```

Comment:
Something feels off about this function using event loop #defines and not default socket type